text/0093-rocksdb-per-region.md

```diff
@@ -0,0 +1,196 @@
+
+Similar to split, merge should also generate new tablet. The purpose is to make it convenient to handle abnormal cases like restarting. Only after the merge command is persisted, should the tablet index in raft db be updated and can source tablet be deleted.
+
+```mermaid
+flowchart TB
+    mem0-->phys2
+    mem1-->phys2
+    phys2-->|hardlink|phys0
+    phys2-->|hardlink|phys1
+    subgraph 2_11
+    mem0["Frozen memtable"]
+    phys0["sst on disk"]
+    end
+    subgraph 3_5
+    mem1["Frozen memtable"]
+    phys1["sst on disk"]
+    end
+    subgraph 3_20
+    mem2["active memtable"]
+    phys2["sst on disk"]
+    end
+```
+
+### Read and write
+
+Most of the read and write design is covered in [dynamic size region][1] RFC. Isolated rocksdb per region can make writes faster as no concurrent conflict and use of prefix hint. If single thread write becomes bottleneck, we can also use unordered apply. Unordered apply is a standalone optimization, I’m not going into details here.
+
+Prototype shows the regression problem of v6.1 is solved in dynamic regions. Dynamic regions have steady throughputs in long running benchmarks like sysbench、tpcc、ycsb. And it also triggers less compactions. 
+
+### Compatibility
+
+#### Upgrade
+
+The storage architecture is very different between v6.1 and dynamic regions. Changes of metadata can be updated every easily during rolling update, however migrating data is nearly impossible. There are two problems:
+- Multiple regions share the same rocksdb in v5.x, split them into tablets takes too much time and can bring much additional writes;
+- There are many small regions in v5.x, which can be more than 100k in a single instance. Upgrading them at once probably OOM.
+
+To solve the problem, let’s introduce a migration stage. Before upgrading, PD should merge small regions into large one. When the region count is small enough, or the size of a region becomes large enough, PD triggers tablet split. After all regions store their data in tablets, the upgrade is finished. PD should provide a switch to users for pausing, resuming, checking the progress of upgrading and controlling its speed. Upgrading is enabled by default, and service should not be interrupted during that time.
```

Comment:
`PD triggers tablet split`------here the tablet split will split the region from single rocksdb into tablets with multirocksdb, right?  It's different with tablet split of either single rocksdb or multi rocksdb.  Probably need some details about it.  

Also during the upgrade, we will have some regions sharing a rocksdb instance, while other regions have dedicated rocksdb instance. It seems to me this state can last very very long. And so back to Shuning's earlier question,  we may support multiple regions sharing single rocksdb as it's the situation in upgrade anyway.

Reply:
> It's different with tablet split of either single rocksdb or multi rocksdb.

Actually it can share the same method as multiple rocksdb. But I will leave it as an implementation decision.

> we may support multiple regions sharing single rocksdb as it's the situation in upgrade anyway.

Yes, supporting *1* mixed storage + *n* isolated storage is possible. But it's a too complicated architecture, and has not much benefits compared to fully isolated solution.

Reply:
`It's different with tablet split of either single rocksdb or multi rocksdb.`
I think here it has some details to worth to discuss. One option is to first move the region from single rocksdb to multi rocksdb and then use multi rocksdb's method for splitting. 
The other option is to read the data of new regions directly and then create a rocksdb instance with that. To me the first option is simpler and better.

Reply:
`Yes, supporting 1 mixed storage + n isolated storage is possible. But it's a too complicated architecture, and has not much benefits compared to fully isolated solution.`
My point is during the upgrade ( I suppose regions are upgraded one by one), we have 1 mixed storage and n isolated storage anyway. And this period can be very long depending on the data size, so we need to implement it anyway as part of the upgrade?

Reply:
> first move the region from single rocksdb to multi rocksdb and then use multi rocksdb's method for splitting.

I don't see the point of moving it to multi rocksdb first. Just use `FreezeAndClone` should be sufficient.

> so we need to implement it anyway as part of the online upgrade?

Not really. If we found the implementation is too complicated, we can still fallback to offline upgrading or primary/sencondary upgrading. Nevertheless, the upgrading mode is not supposed to be performant, it's a temporary state. For example, we may disable certain features in upgrading states. All of this RFC is designed for **physically isolation**.